### ISSUE TITLE
After merge to master: build, version, sign, and upload through Travis

### ISSUE THREAD

**Issue by kennsippell** [medic/medic-android#78]
Creating a public ticket:
https://github.com/medic/medic-projects/issues/2319
https://github.com/medic/medic-projects/issues/3186

I've spent the last few days tinkering. This is a short update with progress, remaining obstacles, and some questions about the approach.

Currently, I have a travis config which builds and uploads release-ready APKs for all flavors. It triggers for new tags in `master` which match the semver schema (`a.b.c`). Builds are optimized, versioned based on the tag, and signed. Should be release-ready, but haven't done any e2e testing.

**Build Times**
The build completes in about ~40-45 minutes. Travis has a timeout of 50 minutes. I've never seen it timeout yet, but we're walking pretty close to the line and will likely cross it if we add another flavor. I'm going to try to create two jobs which both build ~half of the flavors. Or I'm going to whitelist the flavors we actually want to build and release. Seems like ~5 minutes of setup/install, about ~30 minutes to  build, and ~10 minutes (its about 500MB) to upload to GitHub. 

*Warning When we Upload our APK to Google's App Console*
```
Unoptimized APK
Warning:

This APK results in unused code and resources being sent to users. Your app could be smaller if you used the Android App Bundle. By not optimizing your app for device configurations, your app is larger to download and install on users' devices than it needs to be. Larger apps see lower install success rates and take up storage on users' devices.

Resolution:
Use the Android App Bundle to automatically optimize for device configurations, or manage it yourself with multiple APKs.
```

This appears to be some fallout from our change to both `armeabi-v7a` and `arm64-v8a`.  So I've spent some time trying to build bundles instead of APKs. I've got this building, signed, but not quite deploying (one sticky bug remaining). But in Google's App Console I'm getting this error when I try to deploy these bundles: `To upload an Android App Bundle you must be enrolled in App Signing by Google Play.`. I'm unclear if we want to try building multiple APKs, or if we want to try bundling with Google's App Signing. Seems that Google heavily recommends the use of App Signing, but I'm unclear on how that would impact our side loading. @henokgetachew?

**Tests**
I have disabled tests at the moment, but will likely need a new travis job to get that going again also.

**Collect and Gateway**
I'll reproduce this process for `medic-collect` and `medic-gateway` after I completed a PR for this one. I'm wondering if we need x64 builds for those projects also?

**Automated Deploys to App Store**
This is a can of worms I have not yet opened.

# What I'm planning to deliver
* **For QA** - I'm only building based off of changes in master and I'm only building release-ready APKs. So for AT scenarios, my expectation is that you'll continue to build and deploy whatever you need manually `git checkout && make`. For release testing, builds will be available. @newtewt @ngaruko. Does that sound okay?
*  **For Side Loading and New Flavors** - When a build triggers, I'm going to build and deploy every flavor. At this time, I'm not considering any sort of interface or steps that folks can use to trigger a specific build or a specific deployment.
* **Play Store** - Like I said, I haven't touched this. But I'm thinking that the same builds we push to GitHub, we can also push to Play Store as an "alpha" release. Any promotion to production, or promotion to beta releases will need to be done manually.

**Comment by abbyad** [comment]
> **Collect and Gateway**
> I'll reproduce this process for medic-collect and medic-gateway after I completed a PR for this one. I'm wondering if we need x64 builds for those projects also?

Worth checking if they are 64 bit compatible as it would be nice for them to be by Aug 1, 2019. You might find [useful notes about 64-bit applications](https://developer.android.com/distribute/best-practices/develop/64-bit). That said, since those two apps cannot currently be on the Play store due to SMS permissions, it may be less pressing.

**Comment by kennsippell** [comment]
Note to AT: I merged this because 1) the build was broken until it merged, and 2) the MoH Zanzibar team is using v0.4.27, so even if there are bugs, I figure we should commit that build to master for our records.  

Happy to address any issues found in AT in the 3.6 release.

**Comment by newtewt** [comment]
Spot checked a few of the apps and they're pulling down the different logos. Also the unbranded 64 and non 64 is working.
